### PR TITLE
Fix watch stream starvation

### PR DESCRIPTION
**Pull request by serathius**
Ref https://github.com/etcd-io/etcd/issues/15402

Cleaning up serving code and moving grpc server from under http server. Now multiplexing between grpc API and http API will be the same for both TLS and non-TLS config. It will be done by checking HTTP1 (http) vs HTTP2 (grpc).

cc @ahrtr @ptabor

### CODE REVIEW

**Comment by serathius** [comment]
This change breaks curl test :( 
```
curl: (92) HTTP/2 stream 1 was not closed cleanly before end of the underlying stream
```
Marking as draft for debug

File: tests/common/grpc_test.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	g.Go(func() error {
+		lastSize := size
+		for range time.Tick(time.Second) {
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+			}
+			mux.RLock()
+			t.Logf("Generating read load around %.1f MB/s", float64(size-lastSize)/1000/1000)
```
Comment (ptabor) [inline tests/common/grpc_test.go:215]: probably in this statistics would be nice to write how many calls succeeded/deadlined/failed.

File: tests/common/grpc_test.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	if sinceLast > maxDelay && sinceLast > watchResponsePeriod+maxWatchDelay {
+		t.Errorf("Unexpected watch response delayed over allowed threshold %s, delay: unknown", maxWatchDelay)
+		t.Errorf("Test finished while in middle of delayed response, measured delay: %s", sinceLast-watchResponsePeriod)
+		t.Logf("Please increase the test duration to measure delay")
+	} else {
+		t.Logf("Max delay: %s", maxDelay-watchResponsePeriod)
+	}
+}
+
+func fillEtcdWithData(ctx context.Context, c *clientv3.Client, keyCount int, valueSize uint) error {
```
Comment (ptabor) [inline tests/common/grpc_test.go:166]: This sounds like a shared method in the testing framework.

File: tests/common/grpc_test.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+							return err
+						}
+					}
+					time.Sleep(watchResponsePeriod)
+				}
+			})
+			continuouslyExecuteGetAll(ctx, t, &g, c)
+			validateWatchDelay(t, c.Watch(ctx, "key"))
+			err := g.Wait()
+			if err != nil {
```
Comment (ptabor) [inline tests/common/grpc_test.go:134]: require.NoError(g.Wait()) ?

File: server/embed/serve.go [hunk 4]
```diff
@@ -109,113 +111,78 @@ func (sctx *serveCtx) serve(
...
-		srv := &http.Server{
-			Handler:   createAccessController(sctx.lg, s, httpmux),
-			TLSConfig: tlscfg,
-			ErrorLog:  logger, // do not log user error
-		}
-		if err := configureHttpServer(srv, s.Cfg); err != nil {
-			sctx.lg.Error("Configure https server failed", zap.Error(err))
+	listener := sctx.l
+	if sctx.secure {
+		listener, err = transport.NewTLSListener(listener, tlsinfo)
```
Comment (ptabor) [inline server/embed/serve.go:171]: Are we covering here /healthz and /metrics endpoints with mTLS that used to be unprotected previously ?

https://github.com/etcd-io/etcd/issues/11993
https://github.com/etcd-io/etcd/pull/12370#issuecomment-714676715

File: tests/common/grpc_test.go (hunk 1 above)
Comment (serathius) [reply tests/common/grpc_test.go:215]: All calls succeed, any failure would mark test as failed.

File: tests/framework/e2e/e2e.go [hunk 5]
```diff
@@ -51,6 +51,7 @@ func (e e2eRunner) NewCluster(ctx context.Context, t testing.TB, opts ...config.
 		WithStrictReconfigCheck(cfg.StrictReconfigCheck),
 		WithAuthTokenOpts(cfg.AuthToken),
 		WithSnapshotCount(cfg.SnapshotCount),
+		WithWatchProcessNotifyInterval(cfg.WatchProgressNotifyInterval),
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/framework/e2e/e2e.go:54]: Yes, this change comes with 3 dedicated tests for all types of watch response.

File: server/embed/serve.go [hunk 6]
```diff
@@ -109,113 +111,78 @@ func (sctx *serveCtx) serve(
...
+		if err != nil {
 			return err
 		}
-		go func() { errHandler(srv.Serve(tlsl)) }()
+	}
+	m := cmux.New(listener)
+	go func() { errHandler(srv.Serve(m.Match(cmux.HTTP1()))) }()
+	go func() { errHandler(gs.Serve(m.Match(cmux.HTTP2()))) }()
```
[the remark this answers is not in the document]
Comment (serathius) [reply server/embed/serve.go:178]: I think what you suggest will be better, however we need to validate if it breaks non-TLS in any way.

File: tests/common/grpc_test.go (hunk 2 above)
Comment (serathius) [reply tests/common/grpc_test.go:166]: Yep, however let's keep refactor only when needed.

File: tests/common/grpc_test.go (hunk 3 above)
Comment (serathius) [reply tests/common/grpc_test.go:134]: Done.

File: server/embed/serve.go (hunk 4 above)
Comment (serathius) [reply server/embed/serve.go:171]: Don't think so. There is no healthz or metrics handler that I'm touching there. The only existing handler in serve was under TLS and will be under TLS.

**Comment by serathius** [comment]
Closing for https://github.com/etcd-io/etcd/pull/15431

**Pull request closed without merging**